File: ios/Game6561/Data/GamePreferences.swift
import Foundation

class GamePreferences {
    private let defaults = UserDefaults.standard

    private enum Keys {
        static let grid = "grid"
        static let score = "score"
        static let best = "best"
        static let moves = "moves"
        static let timer = "timer"
        static let gameWon = "gameWon"
        static let gameOver = "gameOver"
        static let gamesPlayed = "gamesPlayed"
        static let gamesWon = "gamesWon"
        static let totalScore = "totalScore"
        static let bestScore = "bestScore"
        static let totalMoves = "totalMoves"
        static let maxCombo = "maxCombo"
        static let sound = "sound"
        static let theme = "theme"
        static let tutorial = "tutorial"
    }

    func saveGameState(_ state: GameState) {
        defaults.set(encodeGrid(state.grid), forKey: Keys.grid)
        defaults.set(state.score, forKey: Keys.score)
        defaults.set(state.best, forKey: Keys.best)
        defaults.set(state.moves, forKey: Keys.moves)
        defaults.set(state.gameTimer, forKey: Keys.timer)
        defaults.set(state.gameWon, forKey: Keys.gameWon)
        defaults.set(state.gameOver, forKey: Keys.gameOver)
    }

    func loadGameState() -> GameState? {
        guard let gridStr = defaults.string(forKey: Keys.grid) else { return nil }
        return GameState(
            grid: decodeGrid(gridStr),
            score: defaults.integer(forKey: Keys.score),
            best: defaults.integer(forKey: Keys.best),
            moves: defaults.integer(forKey: Keys.moves),
            gameTimer: defaults.integer(forKey: Keys.timer),
            gameWon: defaults.bool(forKey: Keys.gameWon),
            gameOver: defaults.bool(forKey: Keys.gameOver)
        )
    }

    func clearGameState() {
        for key in [Keys.grid, Keys.score, Keys.best, Keys.moves,
                     Keys.timer, Keys.gameWon, Keys.gameOver] {
            defaults.removeObject(forKey: key)
        }
    }

    var statistics: GameStatistics {
        GameStatistics(
            gamesPlayed: defaults.integer(forKey: Keys.gamesPlayed),
            gamesWon: defaults.integer(forKey: Keys.gamesWon),
            totalScore: defaults.integer(forKey: Keys.totalScore),
            bestScore: defaults.integer(forKey: Keys.bestScore),
            totalMoves: defaults.integer(forKey: Keys.totalMoves),
            maxCombo: defaults.integer(forKey: Keys.maxCombo)
        )
    }

    func recordGame(won: Bool, finalScore: Int, finalMoves: Int, finalMaxCombo: Int) -> GameStatistics {
        var stats = self.statistics
        stats.gamesPlayed += 1
        if won { stats.gamesWon += 1 }
        stats.totalScore += finalScore
        stats.bestScore = max(stats.bestScore, finalScore)
        stats.totalMoves += finalMoves
        stats.maxCombo = max(stats.maxCombo, finalMaxCombo)

        defaults.set(stats.gamesPlayed, forKey: Keys.gamesPlayed)
        defaults.set(stats.gamesWon, forKey: Keys.gamesWon)
        defaults.set(stats.totalScore, forKey: Keys.totalScore)
        defaults.set(stats.bestScore, forKey: Keys.bestScore)
        defaults.set(stats.totalMoves, forKey: Keys.totalMoves)
        defaults.set(stats.maxCombo, forKey: Keys.maxCombo)
        return stats
    }

    var isSoundEnabled: Bool {
        get { defaults.object(forKey: Keys.sound) == nil ? true : defaults.bool(forKey: Keys.sound) }
        set { defaults.set(newValue, forKey: Keys.sound) }
    }

    var isDarkTheme: Bool {
        get { defaults.string(forKey: Keys.theme) == "dark" }
        set { defaults.set(newValue ? "dark" : "light", forKey: Keys.theme) }
    }

    var hasSeenTutorial: Bool {
        get { defaults.bool(forKey: Keys.tutorial) }
        set { defaults.set(newValue, forKey: Keys.tutorial) }
    }

    func resetTutorial() {
        defaults.removeObject(forKey: Keys.tutorial)
    }

    // MARK: - Encoding

    private func encodeGrid(_ grid: [[Int]]) -> String {
        grid.flatMap { $0 }.map(String.init).joined(separator: ",")
    }

    private func decodeGrid(_ str: String) -> [[Int]] {
        let values = str.split(separator: ",").compactMap { Int($0) }
        var grid = Array(repeating: Array(repeating: 0, count: gridSize), count: gridSize)
        for r in 0..<gridSize {
            for c in 0..<gridSize {
                let idx = r * gridSize + c
                if idx < values.count { grid[r][c] = values[idx] }
            }
        }
        return grid
    }
}

File: android/app/src/main/java/com/game6561/app/data/GamePreferences.kt
package com.game6561.app.data

import android.content.Context
import com.game6561.app.game.GameState
import com.game6561.app.game.GameStatistics
import com.game6561.app.game.GRID_SIZE

class GamePreferences(context: Context) {

    private val prefs = context.getSharedPreferences("game6561", Context.MODE_PRIVATE)

    fun saveGameState(state: GameState) {
        prefs.edit()
            .putString("grid", encodeGrid(state.grid))
            .putInt("score", state.score)
            .putInt("best", state.best)
            .putInt("moves", state.moves)
            .putInt("timer", state.gameTimer)
            .putBoolean("gameWon", state.gameWon)
            .putBoolean("gameOver", state.gameOver)
            .putInt("combo", state.combo)
            .putInt("totalMerges", state.totalMerges)
            .putInt("maxCombo", state.maxCombo)
            .apply()
    }

    fun loadGameState(): GameState? {
        val gridStr = prefs.getString("grid", null) ?: return null
        return GameState(
            grid = decodeGrid(gridStr),
            score = prefs.getInt("score", 0),
            best = prefs.getInt("best", 0),
            moves = prefs.getInt("moves", 0),
            gameTimer = prefs.getInt("timer", 0),
            gameWon = prefs.getBoolean("gameWon", false),
            gameOver = prefs.getBoolean("gameOver", false),
            combo = prefs.getInt("combo", 0),
            totalMerges = prefs.getInt("totalMerges", 0),
            maxCombo = prefs.getInt("maxCombo", 0)
        )
    }

    fun clearGameState() {
        prefs.edit()
            .remove("grid")
            .remove("score")
            .remove("best")
            .remove("moves")
            .remove("timer")
            .remove("gameWon")
            .remove("gameOver")
            .remove("combo")
            .remove("totalMerges")
            .remove("maxCombo")
            .apply()
    }

    fun getStatistics(): GameStatistics = GameStatistics(
        gamesPlayed = prefs.getInt("gamesPlayed", 0),
        gamesWon = prefs.getInt("gamesWon", 0),
        totalScore = prefs.getInt("totalScore", 0),
        bestScore = prefs.getInt("bestScore", 0),
        totalMoves = prefs.getInt("totalMoves", 0),
        maxCombo = prefs.getInt("maxCombo", 0)
    )

    fun recordGame(won: Boolean, finalScore: Int, finalMoves: Int, maxCombo: Int): GameStatistics {
        val stats = getStatistics()
        val newStats = GameStatistics(
            gamesPlayed = stats.gamesPlayed + 1,
            gamesWon = stats.gamesWon + (if (won) 1 else 0),
            totalScore = stats.totalScore + finalScore,
            bestScore = maxOf(stats.bestScore, finalScore),
            totalMoves = stats.totalMoves + finalMoves,
            maxCombo = maxOf(stats.maxCombo, maxCombo)
        )
        saveStatistics(newStats)
        return newStats
    }

    private fun saveStatistics(stats: GameStatistics) {
        prefs.edit()
            .putInt("gamesPlayed", stats.gamesPlayed)
            .putInt("gamesWon", stats.gamesWon)
            .putInt("totalScore", stats.totalScore)
            .putInt("bestScore", stats.bestScore)
            .putInt("totalMoves", stats.totalMoves)
            .putInt("maxCombo", stats.maxCombo)
            .apply()
    }

    // Settings
    fun isSoundEnabled(): Boolean = prefs.getBoolean("sound", true)
    fun setSoundEnabled(enabled: Boolean) = prefs.edit().putBoolean("sound", enabled).apply()

    fun getTheme(): String = prefs.getString("theme", "light") ?: "light"
    fun setTheme(theme: String) = prefs.edit().putString("theme", theme).apply()

    fun hasSeenTutorial(): Boolean = prefs.getBoolean("tutorial", false)
    fun markTutorialSeen() = prefs.edit().putBoolean("tutorial", true).apply()
    fun resetTutorial() = prefs.edit().remove("tutorial").apply()

    private fun encodeGrid(grid: Array<IntArray>): String {
        val sb = StringBuilder()
        for (r in 0 until GRID_SIZE) {
            for (c in 0 until GRID_SIZE) {
                if (sb.isNotEmpty()) sb.append(',')
                sb.append(grid[r][c])
            }
        }
        return sb.toString()
    }

    private fun decodeGrid(str: String): Array<IntArray> {
        val values = str.split(',').map { it.toIntOrNull() ?: 0 }
        val grid = Array(GRID_SIZE) { IntArray(GRID_SIZE) }
        for (r in 0 until GRID_SIZE) {
            for (c in 0 until GRID_SIZE) {
                val idx = r * GRID_SIZE + c
                if (idx < values.size) grid[r][c] = values[idx]
            }
        }
        return grid
    }
}
